lib/internal/test_runner/runner.js

```diff
@@ -446,21 +456,43 @@ function run(options) {
   }
 
   const root = createTestTree({ concurrency, timeout, signal });
-  const testFiles = files ?? createTestFileList();
+  let testFiles = files ?? createTestFileList();
 
   let postRun = () => root.postRun();
   let filesWatcher;
   if (watch) {
+    // Overriding testFiles so we won't try to consume again the stream
+    testFiles = ArrayIsArray(testFiles) ? testFiles : testFiles.toArray();
+
+    // TODO - should not pass a promise
     filesWatcher = watchFiles(testFiles, root, inspectPort, signal, testNamePatterns);
     postRun = undefined;
   }
+
+  const runSingleFile = (path) => {
+    const subtest = runTestFile(path, root, inspectPort, filesWatcher, testNamePatterns);
+    filesWatcher?.runningSubtests.set(path, subtest);
+    return subtest;
+  };
+
   const runFiles = () => {
     root.harness.bootstrapComplete = true;
-    return SafePromiseAllSettledReturnVoid(testFiles, (path) => {
-      const subtest = runTestFile(path, root, inspectPort, filesWatcher, testNamePatterns);
-      filesWatcher?.runningSubtests.set(path, subtest);
-      return subtest;
-    });
+
+    if (ArrayIsArray(testFiles)) {
+      testFiles = PromiseResolve(testFiles);
+    }
+
+    // TODO - find better way to check if promise
+    if (typeof testFiles.then === 'function') {
```

Comment:
if it's not a promise, it's an array, right? could this check be inverted, so it can use IsArray, and then PromisePrototypeThen can be used unconditionally, and if it's not a promise, it'll throw?

(or if it's not an array and i'm just confused because it has `.map`, presumably there's another way to check whatever it is)

Reply:
`files` can be array when provided by the user

it can be:
- array: when provided by the user
- stream: when we find the tests
- promise: when we find the tests but we are in watch mode

Reply:
ok, so IsArray works for array, and surely we have a way to brand check node's own streams?

Reply:
do we need to? checked array and promise...

Reply:
right, but there isn't a brand check for Promise, and this TODO comment references it.

Therefore, if you flip the conditions so you check the two other options, then you can just assume it's a promise.

Reply:
gotcha, thanks

on another note, is there a way not to get a bazillion emails when the build cancels due to a new commit being pushed? it's really annoying!

Reply:
I think the common interface should be anything that is async iterable. (arrays are, so are readable streams, so only watch mode needs to be adjusted to use an async iterable - that should not be hard since it uses an event emitter)